src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py

```diff
@@ -0,0 +1,196 @@
+        }
+
+    def _rotate_profiles(self, current_time_slot: Optional[DateTime] = None):
+        self._water_return_temp_C.read_or_rotate_profiles()
+        self._water_supply_temp_C.read_or_rotate_profiles()
+        self._dh_water_flow_m3.read_or_rotate_profiles()
+        self.state.delete_past_state_values(current_time_slot)
+
+    def _calc_energy_to_buy_maximum(self, time_slot: DateTime) -> float:
+        max_energy_consumption = self._storage_temp_to_energy(self._max_temp_C, time_slot)
+        assert max_energy_consumption > -FLOATING_POINT_TOLERANCE
+        return min(self._max_energy_consumption_kWh, max_energy_consumption)
+
+    def _calc_energy_to_buy_minimum(self, time_slot: DateTime) -> float:
+        max_energy_consumption = self._storage_temp_to_energy(
+            self.state.get_storage_temp_C(time_slot), time_slot)
+        assert max_energy_consumption > -FLOATING_POINT_TOLERANCE
+        return min(self._max_energy_consumption_kWh, max_energy_consumption)
+
+    def _calc_temp_decrease_K(self, time_slot: DateTime) -> float:
+        dh_supply_temp = self._water_supply_temp_C.profile[time_slot]
+        dh_return_temp = self._water_return_temp_C.profile[time_slot]
+        m_m3 = self._dh_water_flow_m3.profile[time_slot]
+        m_kg_per_sec = m_m3 * 1000 / 3600
+        q_out = m_kg_per_sec * WATER_SPECIFIC_TEMPERATURE * (dh_supply_temp - dh_return_temp)
+        temp_differential_per_sec = -q_out / (
+                WATER_DENSITY * WATER_SPECIFIC_TEMPERATURE * self._tank_volume_l)
+        temp_decrease = temp_differential_per_sec * GlobalConfig.slot_length.total_seconds()
+        assert temp_decrease <= 0.0
+        return abs(temp_decrease)
+
+    def _calc_temp_increase_K(self, time_slot: DateTime, energy_kWh: float) -> float:
+        storage_temp = self._energy_to_storage_temp(energy_kWh, time_slot)
+        current_storage_temp = self.state.get_storage_temp_C(time_slot)
+        if storage_temp <= current_storage_temp:
+            return 0
+        return storage_temp - current_storage_temp
+
+    def _storage_temp_to_energy(self, storage_temp: float, time_slot: DateTime) -> float:
+        if not self._min_temp_C < storage_temp < self._max_temp_C:
+            logger.info(
+                "Storage temp %s cannot exceed min (%s) / max (%s) tank temperatures.",
+                storage_temp, self._min_temp_C, self._max_temp_C)
+            storage_temp = max(min(storage_temp, self._max_temp_C), self._min_temp_C)
+
+        dh_supply_temp = self._water_supply_temp_C.profile[time_slot]
+        dh_return_temp = self._water_return_temp_C.profile[time_slot]
+        current_storage_temp = self.state.get_storage_temp_C(time_slot)
+        m_m3 = self._dh_water_flow_m3.profile[time_slot]
+        m_kg_per_sec = m_m3 * 1000 / 3600
+        q_out = m_kg_per_sec * WATER_SPECIFIC_TEMPERATURE * (dh_supply_temp - dh_return_temp)
+
+        if m_kg_per_sec <= FLOATING_POINT_TOLERANCE:
+            return 0.
```

Comment:
NIT: could be placed directly after line 113.

Reply:
Fixed